blog-post/blog-post.md

```diff
@@ -0,0 +1,738 @@
+    function mul(lhs : T, rhs : T) -> T;
+}
+```
+
+Instead of the concrete `wmul` function that we defined above for our `wad` fixed point type, it
+would be more idiomatic to defined an instance (a.k.a `impl`) of the `Mul` type class for `wad`
+(giving us a uniform syntax for multiplication across all types, and allowing us to use our `wad`
+type in functions that are generic over any instance of the `wad` type class):
+
+```solidity
+instance wad:Mul {
+    function mul(lhs : wad, rhs : wad) -> wad {
+        return wmul(lhs, rhs);
+    }
+}
+```
+
+If we want to write a function that can accept any type that is an instance of `Mul` we need to add
+a constraint to the signature:
+
+```solidity
+forall T . T:Mul => function square(val : T) -> T {
+    return Mul.mul(val, val);
+}
+```
+
+Simple wrapper types like `wad` are very common, and one type class that can be particularly helpful
+when working with them is `Typedef`:
+
+```solidity
+forall T U . class T:Typedef(U) {
+    function abs(x:T) -> U;
+    function rep(x:U) -> T;
+}
+```
+
+The `abs` (abstraction) and `rep` (representation) functions let us move between wrapper types and
+their underlying types in a generic way and without the syntactic noise of having to introduce
+pattern matching every time we want to unwrap a type. The instance for `wad` would look like this:
+
+```solidity
+instance wad:Typedef(uint256) {
+    function abs(u: uint256) -> wad {
+        return wad(u);
+    }
+
+    function rep(x: wad) -> uint256 {
+        match x {
+        | wad(u) => return u;
+        }
+    }
+}
+```
+
+Note that `U` parameter in the above `Typedef` definition is "weak": its value is uniquely
+determined by the value of the `T` parameter. If you are familiar with Haskell or Rust, this is
+effectively an associated type (although for any type system nerds reading, we implement it using a
+restricted form of functional dependencies). To put it more plainly, we can only implement a single
+instance of `Typedef` for `wad`: the compiler would not allow us to implement both
+`wad:Typedef(uint256)` and `wad:Typedef(uint128)`. This restriction makes type inference much more
+predictable and reliable by sidestepping many of the potential ambiguities inherent to full
+multi-parameter type classes.
+
+For a real world example of how generics and type class constraints can be used to eliminate
+boilerplate or repetitive code, compare the combinatorial explosion of overloads required for the
+[`console.log` implementation](https://github.com/foundry-rs/forge-std/blob/master/src/console.sol)
+in `forge-std` to the following generic Core Solidity function that covers the functionality of all
+the single argument overloads from the original library. The `word` type used in this implementation
+is a low level type that represents a Yul variable, and is the only type that can be passed into or
+out of assembly blocks.
+
+```solidity
+forall T . T:ABIEncode => function log(val : T) {
```

Comment:
in the future it would be cool to show how many LoC we save by having this vs the current forge implementation